Fix plugin engine example

caitlinross opened the pull request:
Fixes #3492. For some reason a `PerformPuts()` was removed that causes the example to break. The example is used in one of the install tests, but it wasn't returning failure states when it happens so that's why we missed in CI. That's been fixed as well. It also turns out that that has been failing for awhile due to not being able to find the plugin libs, so that should be fixed now too.

caitlinross commented:
@vicentebolea will you review? Also not sure how to fix the failure on windows in this case. 

```
 723: 1: [Tue Feb 21 22:50:32 2023] [ADIOS2 WARNING] <Plugins> <PluginManager> <LoadPlugin> : The plugin PluginEngineWrite could not be loaded. Double check ADIOS2_PLUGIN_PATH or plugin library name is correct.
 723: 1: ADIOS2_PLUGIN_PATH: D:/a/ADIOS2/ADIOS2/win2019-vs2019-mpi/Testing/install/cmake/EnginePlugin
```

In `testing/install/EnginePlugin/CMakeLists.txt` I set `ADIOS2_PLUGIN_PATH` environment variable.  I'm assuming this doesn't work on windows because of the slashes, but I set it using `${CMAKE_CURRENT_BINARY_DIR}` as opposed to something hardcoded, so I don't understand why CMake wouldn't provide a valid windows path in that case. Any ideas on how to fix?

vicentebolea commented:
> @vicentebolea will you review? Also not sure how to fix the failure on windows in this case.
> 
> ```
>  723: 1: [Tue Feb 21 22:50:32 2023] [ADIOS2 WARNING] <Plugins> <PluginManager> <LoadPlugin> : The plugin PluginEngineWrite could not be loaded. Double check ADIOS2_PLUGIN_PATH or plugin library name is correct.
>  723: 1: ADIOS2_PLUGIN_PATH: D:/a/ADIOS2/ADIOS2/win2019-vs2019-mpi/Testing/install/cmake/EnginePlugin
> ```
> 
> In `testing/install/EnginePlugin/CMakeLists.txt` I set `ADIOS2_PLUGIN_PATH` environment variable. I'm assuming this doesn't work on windows because of the slashes, but I set it using `${CMAKE_CURRENT_BINARY_DIR}` as opposed to something hardcoded, so I don't understand why CMake wouldn't provide a valid windows path in that case. Any ideas on how to fix?

I think that Cmake is ok with forward slashes even in Windows as long as it is used internally. But that ENV var will be read by a Powershell script or the C/C++ stdlib in Windows which will not accept that path. 

Maybe this cmake function will help to transform the path to _native_, I found it in the file cmake module:

```
file(TO_NATIVE_PATH "<path>" <variable>)
```

vicentebolea reviewed (commented):
Oops I wrote the comments but somehow forgot to press the submit review. Ill have a look to see what can be done for the windows failures

caitlinross commented on examples/plugins/operator/examplePluginOperator_write.cpp:
```diff
@@ -90,5 +92,5 @@ int main(int argc, char *argv[])
         std::cout << e.what() << "\n";
```
I'll just leave it as is. It's just example code (that also gets used in a couple of tests).

Thanks on the cmake function and explanation. That looks like what I need to use!

caitlinross commented:
@vicentebolea So that cmake command appears to have worked, but it's not finding the plugin libs in that directory. Do the built libraries get put in a different directory on Windows? I don't have a Windows system to test this out myself (and haven't really used Windows in 10+ years 😅 ).

vicentebolea commented:
@caitlinross I made some changes to the PR; it appears to be something with the dynamicBinary loading dll files. Not sure, lets see the CI once it finishes

vicentebolea commented:
@caitlinross after some changes finally the DynamicBinder can find the plugin file, however, for some reason there is an exception while loading it. Let me know if you need more help after your return from vacations  :)

caitlinross commented on source/adios2/helper/adiosDynamicBinder.cpp:
```diff
@@ -46,7 +46,6 @@ DynamicBinder::DynamicBinder(std::string libName, std::string libPath)
 #endif
 
     std::vector<std::string> libSuffixes;
-    libSuffixes.emplace_back("");
```
So this shouldn't be removed. It's supposed to be possible that you could tell ADIOS the full library name (e.g. `libMyPlugin.so` instead of just `MyPlugin`), so without this line ADIOS would fail in this situation.

vicentebolea replied:
Oh I see, that makes sense

caitlinross commented:
@vicentebolea It looks like I have the windows error fixed now. I also cleaned up the fixup commits and added some notes to the plugin docs about needing to export symbols. So I think it's ready unless you have any other changes?

caitlinross commented:
Actually one more question, do you want this in release?

vicentebolea commented:
> Actually one more question, do you want this in release?

Yes that's a good idea.

caitlinross commented:
@vicentebolea since I changed the target branch to release, I need you to approve again.
